std_execution.bs

```diff
@@ -1243,6 +1244,133 @@ talk about the semantics of some of the algorithms proposed in this paper.
+    queue of work is run then the thread may wait for an indefinite amount of
+    time. If all thread of the thread-pool are simultaneously performing such
+    blocking operations then deadlock can result.
+
+There are also minor variations on each of these choices. For example:
+
+4.  A variation of (1): Call `std::terminate` if an eager sender is destructed
+    without joining it. This is the approach that `std::thread` destructor
+    takes.
+
+5.  A variation of (2): Request cancellation of the operation before detaching.
+    This reduces the chances of operations continuing to run indefinitely in the
+    background once they have been detached but does not solve the
+    lifetime- or shutdown-related challenges.
+
+6.  A variation of (3): Request cancellation of the operation before blocking on
+    its completion. This is the strategy that `std::jthread` uses for its
+    destructor. It reduces the risk of deadlock but does not eliminate it.
+
+### Eager senders complicate algorithm implementations ### {#design-lazy-algorithms-complexity}
+
+Algorithms that can assume they are operating on senders with strictly lazy
+semantics are able to make certain optimizations that are not available if
+senders can be potentially eager. With lazy senders, an algorithm can safely
+assume that a call to `execution::start` on an operation state strictly happens
+before any receiver methods are called. This frees the algorithm from needing to
+resolve potential race conditions between the start and completion of the work.
+
+When an algorithm needs to deal with potentially eager senders, the potential
+race conditions can be resolved one of several ways, none of which is desirable:
+
+1.  Assume the worst and implement the algorithm defensively, assuming all
+    senders are eager. This obviously has overheads both at runtime and in
+    algorithm complexity. Resolving race conditions is hard.
+
+2.  Require senders to declare whether they are eager or not with a query.
+    Algorithms can then implement two different implementation strategies, one
+    for strictly lazy senders and one for potentially eager senders. This
+    addresses the performance problem of (1) while compounding the complexity
+    problem.
+
+3.  Push the problem back onto eager senders by requiring them to never call a
+    receiver's `set_*` completion until after `execution::start` has been called
+    on the operation state. That puts the onus of dealing with the race
+    condition on the sender author, which is arguably where it belongs. But it
+    has the negative effect of delaying the completion of the operation. Even
+    when the work is done and the continuation available, the sender has to
+    withold the completion, possibly indefinitely.
+
+Of these options, (3) seems the least bad but would seem to run contrary to the
+<i>raison d'être</i> of eager algorithms, which ostensibly is to make the
+results available sooner.
+
```

Comment:
in the GPU space the argument seems to be that eager lets them start faster, not complete faster, yes?

Reply:
I can only imagine they want the work to start faster so that it finishes faster, no?

Reply:
What I heard was that they had limits in getting the data to the GPU and needed that to start before the work was actually started.